src/mscorlib/shared/System/Double.cs

```diff
@@ -238,6 +238,13 @@ public unsafe override int GetHashCode()
                 // Ensure that 0 and -0 have the same hash code
                 return 0;
             }
+
+            if (IsNaN(d))
+            {
+                // Ensure that all NaNs have the same hash code
+                return 1;
```

Comment:
This should just be `d = double.NaN;`, then the subsequent code will do the right thing. We don't want to use 1, as that will cause definite conflicts between NaN and `4.9406564584124654E-324`.